gnovm/stdlibs/std/emit_event.go

```diff
@@ -0,0 +1,77 @@
+package std
+
+// ref: https://github.com/gnolang/gno/pull/853
+
+import (
+	"encoding/json"
+	"errors"
+
+	gno "github.com/gnolang/gno/gnovm/pkg/gnolang"
+	abci "github.com/gnolang/gno/tm2/pkg/bft/abci/types"
+)
+
+func X_emit(m *gno.Machine, typ string, attrs []string) {
+	eventAttrs, err := attrKeysAndValues(attrs)
+	if err != nil {
+		m.Panic(typedString(err.Error()))
+	}
+
+	pkgPath := CurrentRealmPath(m)
+	fnIdent := getPrevFunctionNameFromTarget(m, "Emit")
+	timestamp := getTimestamp(m)
+
+	evtstr := NewGnoEventString(typ, pkgPath, fnIdent, timestamp, eventAttrs...)
+	ctx := m.Context.(ExecContext)
+	ctx.EventLogger.EmitEvent(evtstr)
+}
+
+func attrKeysAndValues(attrs []string) ([]gnoEventAttribute, error) {
+	attrLen := len(attrs)
+	if attrLen%2 != 0 {
+		return nil, errors.New("odd number of attributes. cannot create key-value pairs")
+	}
+	eventAttrs := make([]gnoEventAttribute, attrLen/2)
+	for i := 0; i < attrLen-1; i += 2 {
+		eventAttrs[i/2] = gnoEventAttribute{
+			Key:   attrs[i],
+			Value: attrs[i+1],
+		}
+	}
+	return eventAttrs, nil
+}
+
+type temp gnoEvent
+
+func NewGnoEventString(eventType, pkgPath, ident string, timestamp int64, attrs ...gnoEventAttribute) abci.EventString {
+	evt := newGnoEvent(eventType, pkgPath, ident, timestamp, attrs...)
+
+	jsonRes, err := json.Marshal(map[string]temp{evt.Type: temp(*evt)})
+	if err != nil {
+		panic(err)
+	}
+
+	return abci.EventString(jsonRes)
+}
+
+type gnoEvent struct {
+	Type       string              `json:"-"`
+	PkgPath    string              `json:"pkg_path"`
+	Identifier string              `json:"identifier"`
+	Timestamp  int64               `json:"timestamp"`
```

Comment:
Why do we want Timestamp?

Reply:
In the past, there was a discussion about adding timestamps and this were added at that time.

https://github.com/gnolang/gno/pull/1653#discussion_r1543117945

Reply:
I believe it's a duplicate. We already have the timestamp and block height, so we can easily return those values. Can someone confirm?

PS: @notJoon, could you please verbally confirm when you have resolved a conversation instead of using the "Resolve conversation" button?

Reply:
The `Timestamp` field was dropped by @notJoon at some point, resolving this